### PR TITLE
misc: pyright CI for diffs

### PR DESCRIPTION
**Pull request by superlopuh**

### CODE REVIEW

**Comment by webmiche** [comment]
I didn't realise there exists an action like this. This looks like a cool thing, once we actually merge the pyright CI, so, IMO, we could just include it in that PR.

**Comment by superlopuh** [comment]
I can't get the reviewdog comment to display inline, if anyone has an idea of how to fix it I'd greatly appreciate it!

 My hope is for pyright issues in the diff to be reported by GitHub, so that reviewers can see if new issues are introduced. The tool I found is:
https://github.com/jakebailey/pyright-action
Which uses reviewdog:
https://github.com/reviewdog/reviewdog

I added a test type error but it's not reported

**Comment by superlopuh** [comment]
I think it's a slightly different thing, we have 600 pyright errors in the repo, so we're quite far off from being able to have a 0 pyright errors policy and have the CI report that. This approach would help us at least have a 0-regression policy, so that new code doesn't add more errors

File: xdsl/dialects/affine.py [hunk 1]
```diff
@@ -24,6 +24,9 @@ class For(Operation):
     body: Region
 
     def verify_(self) -> None:
+
+        a: str = 1
```
[the remark this answers is not in the document]
Comment (superlopuh) [reply xdsl/dialects/affine.py:28]: Wham balam

**Comment by superlopuh** [comment]
Works in main framework source code!

File: tests/test_pattern_rewriter.py [hunk 2]
```diff
@@ -293,6 +293,8 @@ def match_and_rewrite(mod: ModuleOp, rewriter: PatternRewriter):
         new_cst = Constant.from_int_and_width(42, i32)
         rewriter.insert_op_at_pos(new_cst, mod.regions[0].blocks[0], 0)
 
+    a: str = 1
```
Comment (superlopuh) [inline tests/test_pattern_rewriter.py:296]: looks like tests aren't supported, or we'd need to set up pyright to run on tests also, can be done in follow-up PRs

**Comment by webmiche** [comment]
> I think it's a slightly different thing, we have 600 pyright errors in the repo, so we're quite far off from being able to have a 0 pyright errors policy and have the CI report that. This approach would help us at least have a 0-regression policy, so that new code doesn't add more errors

Ah, I see. Yes, I like it :)

**Review by math-fehr — APPROVED** [review]
Wow thanks a lot, that's really really nice!

**Comment by math-fehr** [comment]
(We should not forget to remove the wrong code you added for testing though)

**Review by PapyChacal — APPROVED** [review]
Looks great to me! I would be for doing it on tests too.

**Review by webmiche — APPROVED** [review]

File: tests/test_pattern_rewriter.py (hunk 2 above)
Comment (superlopuh) [inline tests/test_pattern_rewriter.py:296]:
```suggestion
```

File: xdsl/dialects/affine.py [hunk 3]
```diff
@@ -24,6 +24,9 @@ class For(Operation):
     body: Region
 
     def verify_(self) -> None:
+
```
Comment (superlopuh) [inline xdsl/dialects/affine.py:27]:
```suggestion
```

File: xdsl/dialects/affine.py (hunk 1 above)
Comment (superlopuh) [inline xdsl/dialects/affine.py:28]:
```suggestion
```

File: xdsl/dialects/affine.py [hunk 4]
```diff
@@ -24,6 +24,7 @@ class For(Operation):
...
```
Comment (superlopuh) [inline xdsl/dialects/affine.py:27]:
```suggestion
```

File: tests/test_pattern_rewriter.py [hunk 5]
```diff
@@ -293,6 +293,7 @@ def match_and_rewrite(mod: ModuleOp, rewriter: PatternRewriter):
         new_cst = Constant.from_int_and_width(42, i32)
         rewriter.insert_op_at_pos(new_cst, mod.regions[0].blocks[0], 0)
 
+
```
Comment (superlopuh) [inline tests/test_pattern_rewriter.py:296]:
```suggestion
```

**Pull request merged**